src/hotspot/share/utilities/readWriteLock.cpp

```diff
@@ -0,0 +1,147 @@
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/readWriteLock.hpp"
+
+void ReadWriteLock::read_lock(Thread *current) {
+  for (;;) {
+    const int32_t count = Atomic::load_acquire(&_count);
+    if (count < 0) {
+      // Wait until unlocked by writer
+      auto await = [&]() {
+        Locker locker(&_mon);
+        while (Atomic::load_acquire(&_count) < 0) {
+          _mon.wait(0);
+        }
+      };
+      if (current->is_Java_thread()) {
+        ThreadBlockInVM tbivm(JavaThread::cast(current));
+        await();
+      } else {
+        await();
+      }
+      continue;
+    }
+
+    // Increment count
+    if (Atomic::cmpxchg(&_count, count, count + 1) != count) {
+      continue;
+    }
+
+    // Entered critical region
+    return;
+  }
+}
+
+void ReadWriteLock::read_unlock() {
+  for (;;) {
+    const int32_t count = Atomic::load_acquire(&_count);
+
+    if (count > 0) {
+      // No writer in progress, try to decrement reader count.
+      if (Atomic::cmpxchg(&_count, count, count - 1) != count) {
+        continue;
+      }
+    } else {
+      // Writer in progress, try to increment reader count.
+      if (Atomic::cmpxchg(&_count, count, count + 1) != count) {
+        continue;
+      }
+      // If the previous count was -2, then we just incremented it to -1,
+      // and we should signal that all readers have now exited their
+      // critical region and the writer may now block readers.
+      if (count == -2) {
+        Locker locker(&_mon);
+        _mon.notify_all();
+      }
+    }
+    // Exited critical region
+    return;
+  }
+}
+
+void ReadWriteLock::write_lock(Thread *current) {
+  for (;;) {
+    const int32_t count = Atomic::load_acquire(&_count);
+
+    if (count < 0) {
+      // Already has a writer, wait until unlocked
+
+      auto await_writers_exit = [&]() {
+        Locker locker(&_mon);
+        while (Atomic::load_acquire(&_count) < 0) {
+          _mon.wait(0);
+        }
+      };
+
+      if (current->is_Java_thread()) {
+        ThreadBlockInVM tbivm(JavaThread::cast(current));
+        await_writers_exit();
+      } else {
+        await_writers_exit();
+      }
+      continue;
+    }
+
+    // Increment and invert count
+    if (Atomic::cmpxchg(&_count, count, -(count + 1)) != count) {
+      continue;
+    }
+
+    // If the previous count was 0, then we just incremented and inverted
+    // it to -1 and have now blocked readers. Otherwise we wait until all reader
+    // threads have exited the critical region.
+    if (count != 0) {
+      auto await_writer_entry = [&]() {
+        // Wait until all readers exit.
+        Locker locker(&_mon);
+        while (Atomic::load_acquire(&_count) != -1) {
+          _mon.wait(0);
+        }
+      };
+      if (current->is_Java_thread()) {
+        ThreadBlockInVM tbivm(JavaThread::cast(current));
+        await_writer_entry();
+      } else {
+        await_writer_entry();
+      }
+    }
+
+    // Locked.
+    return;
+  }
+}
```

Comment:
I was wondering why a lock for write doesn't hold the underlying
PlatformMonitor across the scope of the mutex region. Something like this
(completely untested)

```
void ReadWriteLock::write_lock(Thread* current) {
  _mon.lock();
  while (true) {
    const int32_t count = Atomic::load_acquire(&_count);
    if (count < 0) {
      // Some other writer is waiting for readers to complete.
      // Approx:
      // while (Atomic::load_acquire(&_count)) _mon.wait();
      await_write_unlock(current);
    } else if (Atomic::cmpxchg(&_count, count, -(count + 1)) == count) {
      // Claimed the write slot, but there might be active readers.
      if (count != 0) {
        // Approx:
        // while (Atomic::load_acquire(&_count) != -1) _mon.wait();
        await_no_active_readers(current);
      }
      return; // return still holding lock
    } // else failed to claim write ownership, so try again.
  }
}

void ReadWriteLock::write_unlock() {
  assert(Atomic::load(&_count) == -1, "invariant");
  Atomic::release_store(&_count, (int32_t)0);
  _mon.notify_all();
  _mon.unlock();
}
```

This seems simpler to me.  Am I missing something?

And looking at this also makes me wonder if `_mon` can just be a normal
HotSpot `Monitor`, which would resolve many of the questions and comments I've
made that are mostly driven by the direct use of `PlatformMonitor`.

Reply:
This lock is essentially unbounded, are there any implications regarding stopping the VM from going into safepoints? I assume not, if we're using `Monitor`.